[RFC 54] Internationalisation reference docs (#6377)
wagtail/wagtail: docs/reference/pages/model_reference.rst

    .. automethod:: get_site_root_paths

``Locale``
==========

The Locale model defines the set of locales that can be used on a site.
Initially, Wagtail will create one ``Locale`` that corresponds to the ``LANGUAGE_CODE`` setting.

When ``WAGTAIL_I18N_ENABLED`` is ``False`` all new pages will inherit their parent's locale.
This is to satisfy the database constraints but also makes makes it easy to switch on internationalisation later.

Changing ``LANGUAGE_CODE``

> It's possible that a site may change it's language code during development. If this happens, we will not update the Locale object or create a new one. If WAGTAIL_I18N_ENABLED is False, pages will continue to inherit their parent's locale even if that doesn't match the current LANGUAGE_CODE.

Reviewer: it's language code -> its language code during

We shouldn't add new locales automatically as there may be cases where changing the language_code of an existing Locale would make more sense.

If a language code is removed from WAGTAIL_CONTENT_LANGUAGES, the corresponding Locale will be filtered out from Locale.objects. This will stop the locale from appearing anywhere in the Wagtail UI and also stop traffic from being routed to it. This Locale will show up with a warning in the Locales UI.
Fields

    language_code (string, unique) - This is a language code taken from the list in WAGTAIL_CONTENT_LANGUAGES. Locales are defined by their language. The language code must be a BCP 47 language tag (the same format used in Django).

Methods

    @classmethod get_default() - This returns the default locale based on the value of the LANGUAGE_CODE setting
    @classmethod get_active() - This returns the currently activated locale
    get_display_name() - This returns the display name that is defined in WAGTAIL_CONTENT_LANGUAGES

``TranslatableMixin``
=====================

> ``TranslatableMixin`` is an abstract model can be applied to any Django model to make them translatable.

Reviewer: make them translatable -> make it translatable

This is already applied to the base ``Page`` model so you don't need to reapply it there.

Fields

    locale Foreign Key to Locale. For pages, this defaults to the locale of the parent page.
    translation_key A UUID that is randomly generated whenever a new model instance is created. This is shared with all translations of that instance so can be used for querying translations.

Both of these fields will have a unique key to prevent something from being translated into the same language twice.

> Note that these fields will exist regardless of whether WAGTAIL_I18N_ENABLED has been set. They shouldn’t get in the way of anything. Neither wagtailtrans nor wagtail-modeltranslation use these field names.

Reviewer: shouldn’t -> shouldn't

    .. automethod:: get_site_root_paths

Locale
======

The Locale model defines the set of locales that can be used on a site.
Each ``Locale`` record corresponds to a language code defined in the ``WAGTAIL_CONTENT_LANGUAGES`` setting.

Wagtail will initially set up one ``Locale`` to act as the default language for all existing content.

> This first locale will automatically pick the value from ``WAGTAIL_CONTENT_LANGUAGES`` that most closely matches the sites primary language code defined in ``LANGUAGE_CODE``.

Reviewer 2: ```suggestion
This first locale will automatically pick the value from ``WAGTAIL_CONTENT_LANGUAGES`` that most closely matches the site primary language code defined in ``LANGUAGE_CODE``.
```

> If the primary language code is changed later, Wagtail will not automatically create a new ``Locale`` record or update an existing one.

Reviewer 2: ```suggestion
If the primary language code is changed later, Wagtail will **not** automatically create a new ``Locale`` record or update an existing one.
```

imho, would be good to emphasise the "not" part otherwise it can could be missed

Before internationalisation is enabled, all pages use this primary ``Locale`` record.

> This is to satisfy the database constraints and makes it easy to switch on internationalisation later.

Reviewer 2: ```suggestion
This is to satisfy the database constraints. And it makes it easier to switch internationalisation on at a later date.
```

Changing ``WAGTAIL_CONTENT_LANGUAGES``
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

Languages can be added or removed from ``WAGTAIL_CONTENT_LANGUAGES`` over time.

Before removing an option from ``WAGTAIL_CONTENT_LANGUAGES``, it's important that the ``Locale``
record is updated to a use a different content language or is deleted.
Any ``Locale`` instances that have invalid content languages are automatically filtered out from all
database queries making them unable to be edited or viewed.

Methods and properties
~~~~~~~~~~~~~~~~~~~~~~

.. class:: Locale
    :noindex:

    .. attribute:: language_code

>         Foo

Reviewer 2: placehoder content or just missed in cleanup?

Author: Oops, yep was a test, Thanks!

    .. automethod:: get_default

    .. automethod:: get_active

    .. automethod:: get_display_name

Translatable Mixin
==================

``TranslatableMixin`` is an abstract model that can be added to any non-page Django model to make it translatable.
Pages already have this mixin added by Wagtail, so no need to add it to them.

Methods and properties
~~~~~~~~~~~~~~~~~~~~~~

> The ``locale`` and ``translation_key`` fields have a unique key to prevent the object being translated into a language more than once.

Reviewer 2: ```suggestion
The ``locale`` and ``translation_key`` fields have a unique key constraint to prevent the object being translated into a language more than once.
```

.. class:: TranslatableMixin
    :noindex:

    .. attribute:: locale

        (Foreign Key to :class:`~wagtail.core.models.Locale`)

         For pages, this defaults to the locale of the parent page.

    .. attribute:: translation_key

        (uuid)

        A UUID that is randomly generated whenever a new model instance is created.
        This is shared with all translations of that instance so can be used for querying translations.

    .. automethod:: get_translations

    .. automethod:: get_translation

    .. automethod:: get_translation_or_none

    .. automethod:: has_translation

    .. automethod:: copy_for_translation

>     .. automethod:: get_translation_model

Reviewer 3: The docstring here says:
> Gets the model which manages the translations for this model. (The model that has the “translation_key” and “locale” fields) Most of the time this would be the current model, but some sites may have intermediate concrete models between wagtailcore.Page and the specfic page model.

The mention of intermediate concrete models doesn't seem right - surely _any_ subclass of Page would be an example of a model where this doesn't return the current model? (Also, best to avoid the word "site" when not talking about Site records - usually I'd say project or installation.)

Perhaps this should be something like: "Typically this would be the current model, but it may be a superclass when multi-table inheritance is in use (as is the case for wagtailcore.Page)."

Author: Fixed, thanks!
